### PR TITLE
Add support to one-to-many @Cascade relationships

### PR DESCRIPTION
**Pull request by mshima**
- Implements backend and angular.

```
application {
  entities *
}

entity Parent {
  name String
}

entity Child {
  name String
}

relationship OneToMany {
  Parent{childs} to @Cascade Child{parent}
}
```

---

Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by mshima** [comment]
@kaidohallik can you have a look?
I've implemented update form using components composition.
Maybe I should split into a html for each component (route, embedded, modal).

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-detail.component.html.ejs [hunk 1]
```diff
@@ -18,7 +18,7 @@
 -%>
 <div class="row justify-content-center">
     <div class="col-8">
-        <div *ngIf="<%= entityInstance %>">
+        <div *ngIf="<%= entityInstance %> != null">
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-detail.component.html.ejs:21]: Why this change?

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-routing.module.ts.ejs [hunk 2]
```diff
@@ -88,6 +88,11 @@ const <%= entityInstance %>Route: Routes = [
         RouterModule.forChild(<%= entityInstance %>Route),
         <%= microservicePrefix %><%= entityAngularName %>Module,
     ],
+    <%_ if (!readOnly) { _%>
+    declarations: [
+        <%= entityAngularName %>RouteUpdateComponent,
+    ],
+    <%_ } _%>
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-routing.module.ts.ejs:95]:
```suggestion
```
Don't declare components in routing module. Use entity module instead.

**Review by kaidohallik — CHANGES_REQUESTED** [review]
> Maybe I should split into a html for each component (route, embedded, modal).

According to Angular style guide: yes, every component should be in separate file.  
https://angular.io/guide/styleguide#application-structure-and-ngmodules states:
```
All content is one asset per file. Each component, service, and pipe is in its own file.
```

If I click on `Create a ne Parent` then I get error:
```
ERROR TypeError: Cannot read property 'forEach' of undefined
    at ParentUpdateComponent.updateForm (parent-update.component.ts:65)
    at ParentUpdateComponent.ngOnInit (parent-update.component.ts:59)
```

For child component currently is generated unused **delete-dialog** and **detail** components (and also **module** is not used, see comment on `files-angular.js` file):
* unused code should not be generated
* in parent **list** and **detail** view should show child information, I think it's not ok to show child information only in update form

Inside child component is generated unused `UpdateModalComponent` component - unused code should not be generated.  
And do we need both `UpdateNestedComponent` and `UpdateComponent`? Currently only `UpdateNestedComponent` is used. I suggest to remove `UpdateComponent` and rename `UpdateNestedComponent` to `UpdateComponent`.

Conclusion from previous comment: seems that we should not generate multiple update components, this seems unnecessary. Generate only one update component per entity and normal entities should be stay unchanged at all.

**Comment by mshima** [comment]
@kaidohallik thanks for you deep review.
This is a pretty much work in progress.

I should have been more specific.
So far I've implemented a components composing approach:
- `entity-update.component.html`
- `EntityUpdateComponent`
- `Entity(Route/Nested/Modal)UpdateComponent` that embeds `EntityUpdateComponent`.

I am wondering if I switch to duplicate:
- extract common code into a partial ejs.
- `entity-(route/nested/modal)-update.component.html`
- `Entity(Route/Nested/Modal)UpdateComponent`

> ```
> All content is one asset per file. Each component, service, and pipe is in its own file.
> ```

Sure, I am keeping in same file so I don't have to remove files for now.

> If I click on `Create a ne Parent` then I get error:

I will have a look, didn't saw the error.

> Conclusion from previous comment: seems that we should not generate multiple update components, this seems unnecessary.

Modal, detail and others components should be used in parent detail view.
Before implementing detail I wanted to discuss what approach to use, just like update component (duplicating or composing).

> Generate only one update component per entity and normal entities should be stay unchanged at all.

If we should keep normal entities unchanged, them the approach is to duplicate components. I am correct?

File: generators/entities-client/files-angular.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * limitations under the License.
+ */
+const { ANGULAR_DIR } = require('../generator-constants');
+
+const angularFiles = {
+    client: [
+        {
+            condition: generator => generator.frontendEmbeddedEntities.length > 0,
+            path: ANGULAR_DIR,
+            templates: ['shared/shared-entity.module.ts'],
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/entities-client/files-angular.js:26]: That is required, a component cannot be imported in two different 'leafs' of a parent component.

Example:
```
@embedded
entity Phone {}

entity Doctor {}

entity Patient {}

relationship OneToMany {
   Doctor to phone
   Patient to phone
}
```

If `DoctorModule` imports `PhoneModule` and `PatientModule` imports `PhoneModule`, angular requires `PhoneModule` to be declared in a common module, in this case `SharedEntityModule`.

**Comment by mshima** [comment]
The approach needs to be changed. Closing.

**Comment by kaidohallik** [comment]
> If we should keep normal entities unchanged, them the approach is to duplicate components. I am correct?

No, duplication is bad. As I saw some components unused then I probably didn't realize the final aim. If inner update component will be used in many places then this is a correct way to go.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs [hunk 4]
```diff
@@ -179,41 +176,59 @@ _%>
...
-            <%_ if (fieldsContainInstant || fieldsContainZonedDateTime) { _%>
+    onSuccess: OnSuccessCallback = (): void => {
+      // do nothing.
+    };
+    onError: OnErrorCallback = (): void => {
+      // do nothing.
+    };
+    previousState = (): void => {
+      // do nothing.
+    };
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs:190]: At the html template side I've tried:
- `(click)="previousState()"`
- `(click)="previousState?.()"`
- `(click)="previousState?()"`

None worked.
It requires a wrapper:
```
  callPreviousState(): void {
    this.previousState?.();
  }
``

[the remark this answers is not in the document]
Comment (kaidohallik) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs:190]: As optional call is quite new TypeScript feature then this is not yet supported in Angular templates. The same workaround is done in `alert-error.component` and `alert.component`.

File: generators/entities-client/files-angular.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaidohallik) [reply generators/entities-client/files-angular.js:26]: It sounds strange. There shouldn't be any difference if import directly or through common shared module. Maybe something else is causing this error?  
JDL sample provided doesn't create `PhoneModule`, can you give example where this problem exists?

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs [hunk 5]
```diff
@@ -122,41 +132,28 @@ type SelectableManyToManyEntity = <%= selectableManyToManyEntities.join(' | ') %
...
 _%>
 
+interface OnSuccessCallback {
+  (response: HttpResponse<I<%= entityAngularName %>>): void;
+}
+
+interface OnErrorCallback {
+  (error: any): void;
+}
+
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs]: Done

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs]: Great.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-routing.module.ts.ejs (hunk 2 above)
Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-routing.module.ts.ejs]: This was a leftover of a different approach .
Done

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-detail.component.html.ejs (hunk 1 above)
Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-detail.component.html.ejs]: Done

File: generators/entities-client/files-angular.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mshima) [reply generators/entities-client/files-angular.js:26]: I forgot the link/error that this approach causes.
I will have to reproduce.

[the remark this answers is not in the document]
Comment (mshima) [reply generators/entities-client/files-angular.js:26]: Seems to be working now.

**Comment by standors** [comment]
is the JDL @Cascade already supported ? It is still not documented for JDL.... 
relationship OneToMany {
  Parent{childs} to @Cascade Child{parent}
}
